### PR TITLE
expression: rewrite builtin function: CONVERT

### PR DESCRIPTION
**Pull request by zz-jason**
to #4080

### CODE REVIEW

File: expression/builtin_string.go [hunk 1]
```diff
@@ -799,49 +799,38 @@ func (c *convertFunctionClass) getFunction(ctx context.Context, args []Expressio
 	if err := c.verifyArgs(args); err != nil {
 		return nil, errors.Trace(err)
 	}
-	sig := &builtinConvertSig{newBaseBuiltinFunc(args, ctx)}
+	bf := newBaseBuiltinFuncWithTp(args, ctx, tpString, tpString, tpString)
```
Comment (XuHuaiyu) [inline expression/builtin_string.go]: we always set charset and collate to utf8 in convert?

``` sql
mysql> select convert("123" using binary);
Field   1:  `convert("123" using binary)`
Catalog:    `def`
Database:   ``
Table:      ``
Org_table:  ``
Type:       VAR_STRING
Collation:  binary (63)
Length:     3
Max_length: 3
Decimals:   31
Flags:      BINARY


+-----------------------------+
| convert("123" using binary) |
+-----------------------------+
| 123                         |
+-----------------------------+
1 row in set (0.00 sec)
```

File: expression/builtin_string.go [hunk 2]
```diff
@@ -799,49 +799,38 @@ func (c *convertFunctionClass) getFunction(ctx context.Context, args []Expressio
...
+	charsetName, isNull, err := b.args[1].EvalString(row, sc)
+	if isNull || err != nil {
+		return "", true, errors.Trace(err)
 	}
 
-	encoding, _ := charset.Lookup(Charset)
+	encoding, _ := charset.Lookup(charsetName)
 	if encoding == nil {
-		return d, errors.Errorf("unknown encoding: %s", Charset)
+		return "", true, errors.Errorf("unknown encoding: %s", charsetName)
```
Comment (jackysp) [inline expression/builtin_string.go]: Which error will mysql return here?

File: expression/builtin_string.go [hunk 3]
```diff
@@ -142,6 +143,8 @@ var (
 	_ builtinFunc = &builtinInstrBinarySig{}
 )
 
+var errUnknownCharacterSet = terror.ClassExpression.New(mysql.ErrUnknownCharacterSet, mysql.MySQLErrName[mysql.ErrUnknownCharacterSet])
```
Comment (XuHuaiyu) [inline expression/builtin_string.go]: move this to expression/ expression.go ?

File: expression/builtin_string.go (hunk 1 above)
Comment (zz-jason) [reply expression/builtin_string.go]: see #4436

Comment (XuHuaiyu) [reply expression/builtin_string.go]: I can't get what you mean?

Comment (zz-jason) [reply expression/builtin_string.go]: Currently, the behavior of function `CONVERT` in TiDB is different from MySQL:
1. As #4436 states: `charset` parameter should be a constant. In `getFunction()` we should set the charset and flag using the constant `charset` parameter.
2. Charsets supported by TiDB may not supported in MySQL, like "866"

**Comment by zz-jason** [comment]
@XuHuaiyu PTAL

File: expression/builtin_string.go (hunk 1 above)
Comment (breeswish) [reply expression/builtin_string.go]: Is convert to binary supported?

Comment (zz-jason) [reply expression/builtin_string.go]: not, it's another issue: Charsets supported by MySQL may not supported in TiDB, like "binary"

Comment (breeswish) [reply expression/builtin_string.go]: Ok

Comment (breeswish) [reply expression/builtin_string.go]: I think a TODO is needed here since after supporting binary charset, we also need to add a binary flag here.

Comment (breeswish) [reply expression/builtin_string.go]: Rest LGTM

Comment (zz-jason) [reply expression/builtin_string.go]: added, PTAL

**Pull request merged**